libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt

```diff
@@ -8,16 +8,114 @@ package org.jetbrains.kotlin.gradle.incremental
+
+        // Use sets to make presence checks faster
+        val unchangedCurrentFiles: Set<File> = unchangedCurrentToPreviousAlignment.keys
+        val unchangedPreviousFiles: Set<File> = unchangedCurrentToPreviousAlignment.values.toSet()
+
+        // We will split the current files into 2 groups:
+        //    1a) Unchanged current files
+        //    1b) Added files
+        // We will split the previous files into 2 groups:
+        //    2a) Unchanged previous files
+        //    2b) Removed files
+        // If the classpath doesn't contain duplicate classes, comparing (1b) with (2b) would be enough.
+        // However, if the classpath contains duplicate classes, comparing (1b) with (2b) would not be enough.
+        // Therefore, to deal with duplicate classes while still being able to compare (1b) with (2b), we will find snapshot files in groups
+        // (1a) and (2a) that have duplicate classes with groups (1b) or (2b) and add them to groups (1b) and (2b). Duplicate classes in
+        // groups (1b) and (2b) will then be handled in a separate step (see ClasspathChangesComputer.getNonDuplicateClassSnapshots).
+        val addedFiles: List<File> = currentClasspathEntrySnapshotFiles.filter { it !in unchangedCurrentFiles }
+        val removedFiles: List<File> = previousClasspathEntrySnapshotFiles.filter { it !in unchangedPreviousFiles }
+
+        val adjustedAddedFiles = addedFiles.toMutableSet()
+        val adjustedRemovedFiles = removedFiles.toMutableSet()
+        unchangedCurrentToPreviousAlignment.forEach { (unchangedCurrentFile, unchangedPreviousFile) ->
+            if (unchangedCurrentFile.containsDuplicatesWith(addedFiles) || unchangedPreviousFile.containsDuplicatesWith(removedFiles)) {
+                adjustedAddedFiles.add(unchangedCurrentFile)
+                adjustedRemovedFiles.add(unchangedPreviousFile)
+            }
+        }
+
+        // Keep the original order of added/removed files as it is important for the handling of duplicate classes.
+        val finalAddedFiles: List<File> = currentClasspathEntrySnapshotFiles.filter { it in adjustedAddedFiles }
+        val finalRemovedFiles: List<File> = previousClasspathEntrySnapshotFiles.filter { it in adjustedRemovedFiles }
+
+        val changedCurrentSnapshot = ClasspathSnapshotSerializer.load(finalAddedFiles)
+        val changedPreviousSnapshot = ClasspathSnapshotSerializer.load(finalRemovedFiles)
+
+        return compute(changedCurrentSnapshot, changedPreviousSnapshot)
+    }
+
+    /**
+     * Maps the unchanged snapshot files of the current build to the unchanged snapshot files of the previous build (selected from all the
+     * snapshot files of the previous build).
+     */
+    private fun alignUnchangedSnapshotFiles(unchangedCurrentSnapshotFiles: List<File>, previousSnapshotFiles: List<File>): Map<File, File> {
+        var index = 0
+        return unchangedCurrentSnapshotFiles.associateWith { unchangedCurrentFile ->
```

Comment:
Can you:
- build 2 maps
    - one for current files Map<FileInfo, List<File>>
    - one for previous files Map<FileInfo, List<File>>

and use them to find the matching size/timestamp?

Reply:
Is it to avoid using the `index` variable? If so, that code would be easier to read, but it might result in incorrect alignments.
 
 For example:
    - Current classpath: `A, B`
    - Previous classpath: `B, A, B`

Gradle reports that `A, B` in the current classpath are unchanged.

The algorithm (without using `index`) would map `A` to `A`, but would map `B` in the current classpath with the first `B` in the previous classpath, which "crosses" the alignment between `A` and `A`, and therefore is incorrect.

Or did you have a different idea?

Reply:
Also, I realized that we can't use timestamps to find candidates as timestamps can change but contents can still be the same. So in the next patch, I'll use only file size to find candidates.

Reply:
My motivation for building a map is to avoid repeatedly querying files for size and timestamp, but to do it only once per file.

WRT not using timestamps, please note that size may also change but Gradle may report that file as unchanged. E.g if the order of entries in a jar changes or timestamp changes or some of the zip entry attributes changes, total jar size may change, but Gradle will ignore all of those.

Reply:
Ah got it, I've uploaded a commit to fix that.

#2 is a good point, I added a code comment:
```
* IMPORTANT: The alignment algorithm must use the same input normalization that is used for snapshot files in the Gradle task.
* Currently, snapshot files are annotated with `@Classpath` and are regular files (not jars), so (only) their contents and order
* matter.
```